File: build.ninja.js
const fs = require("fs");
let out;
let lines = fs.readFileSync("build-template.ninja").toString().split("\n");
out = [];
for (let line of lines) {
   if (line.startsWith(" command = clang ")) continue;
   if (line.startsWith(" command = gcc ")) continue;
   out.push(line + "\n");
}
fs.writeFileSync("build-tcc.ninja", out.join(""));
out = [];
for (let line of lines) {
   if (line.startsWith(" command = tcc ")) continue;
   if (line.startsWith(" command = gcc ")) continue;
   out.push(line + "\n");
}
fs.writeFileSync("build-clang.ninja", out.join(""));
out = [];
for (let line of lines) {
   if (line.startsWith(" command = tcc ")) continue;
   if (line.startsWith(" command = gcc ")) continue;
   if (line.startsWith(" command = clang ")) {
      out.push(line.split("$flags").join("$flags -ggdb") + "\n");
   } else {
      out.push(line + "\n");
   }
}
fs.writeFileSync("build-clang-debug.ninja", out.join(""));
out = [];
for (let line of lines) {
   if (line.startsWith(" command = tcc ")) continue;
   if (line.startsWith(" command = gcc ")) continue;
   if (line.startsWith(" command = clang ")) {
      out.push(line.split("$flags").join("$flags -O3 -flto") + "\n");
   } else {
      out.push(line + "\n");
   }
}
fs.writeFileSync("build-clang-release.ninja", out.join(""));
out = [];
for (let line of lines) {
   if (line.startsWith(" command = tcc ")) continue;
   if (line.startsWith(" command = clang ")) continue;
   out.push(line + "\n");
}
fs.writeFileSync("build-gcc.ninja", out.join(""));
out = [];
for (let line of lines) {
   if (line.startsWith(" command = tcc ")) continue;
   if (line.startsWith(" command = clang ")) continue;
   if (line.startsWith(" command = gcc ")) {
      out.push(line.split("$flags").join("$flags -ggdb") + "\n");
   } else {
      out.push(line + "\n");
   }
}
fs.writeFileSync("build-gcc-debug.ninja", out.join(""));
out = [];
for (let line of lines) {
   if (line.startsWith(" command = tcc ")) continue;
   if (line.startsWith(" command = clang ")) continue;
   if (line.startsWith(" command = gcc ")) {
      out.push(line.split("$flags").join("$flags -O3 -flto") + "\n");
   } else {
      out.push(line + "\n");
   }
}
fs.writeFileSync("build-gcc-release.ninja", out.join(""));

File: build.ninja.py
out = []
for line in open("build-template.ninja", "r"):
    if line.startswith(" command = clang "): continue
    if line.startswith(" command = gcc "): continue
    out.append(line)
f = open("build-tcc.ninja", "wb")
f.write("".join(out).encode("ascii"))
f.close()

out = []
for line in open("build-template.ninja", "r"):
    if line.startswith(" command = tcc "): continue
    if line.startswith(" command = gcc "): continue
    out.append(line)
f = open("build-clang.ninja", "wb")
f.write("".join(out).encode("ascii"))
f.close()

out = []
for line in open("build-template.ninja", "r"):
    if line.startswith(" command = tcc "): continue
    if line.startswith(" command = gcc "): continue
    if line.startswith(" command = clang "):
        line = "$flags -fsanitize=address -ggdb".join(line.split("$flags"))
    out.append(line)
f = open("build-clang-debug.ninja", "wb")
f.write("".join(out).encode("ascii"))
f.close()

out = []
for line in open("build-template.ninja", "r"):
    if line.startswith(" command = tcc "): continue
    if line.startswith(" command = gcc "): continue
    if line.startswith(" command = clang "):
        line = "$flags -O3 -flto".join(line.split("$flags"))
    out.append(line)
f = open("build-clang-release.ninja", "wb")
f.write("".join(out).encode("ascii"))
f.close()

out = []
for line in open("build-template.ninja", "r"):
    if line.startswith(" command = tcc "): continue
    if line.startswith(" command = clang "): continue
    out.append(line)
f = open("build-gcc.ninja", "wb")
f.write("".join(out).encode("ascii"))
f.close()

out = []
for line in open("build-template.ninja", "r"):
    if line.startswith(" command = tcc "): continue
    if line.startswith(" command = clang "): continue
    if line.startswith(" command = gcc "):
        line = "$flags -fsanitize=undefined -ggdb".join(line.split("$flags"))
    out.append(line)
f = open("build-gcc-debug.ninja", "wb")
f.write("".join(out).encode("ascii"))
f.close()

out = []
for line in open("build-template.ninja", "r"):
    if line.startswith(" command = tcc "): continue
    if line.startswith(" command = clang "): continue
    if line.startswith(" command = gcc "):
        line = "$flags -O3 -flto".join(line.split("$flags"))
    out.append(line)
f = open("build-gcc-release.ninja", "wb")
f.write("".join(out).encode("ascii"))
f.close()
